### ISSUE TITLE
IKOS build fails on gcc12

### ISSUE THREAD

**Issue by dunaich** [NASA-SW-VnV/homebrew-core#6]
IKOS 3.1 build failed on gcc12 due to missed `#include <vector>`  in some files:

```
.cache/Homebrew/Logs/ikos/04.make
2023-03-09 10:10:04 +0000

make
install

[  4%] Building CXX object frontend/llvm/CMakeFiles/ikos-pp-lib.dir/src/pass/mark_internal_inline.cpp.o
[  4%] Building CXX object frontend/llvm/CMakeFiles/ikos-pp-lib.dir/src/pass/lower_select.cpp.o
[  4%] Building CXX object frontend/llvm/CMakeFiles/ikos-pp-lib.dir/src/pass/initialize.cpp.o
[  4%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/format/namer.cpp.o
[  4%] Building CXX object frontend/llvm/CMakeFiles/ikos-pp-lib.dir/src/pass/lower_cst_expr.cpp.o
[  4%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/format/dot.cpp.o
[  4%] Generating python/ikos/settings/__init__.py
[  4%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/format/text.cpp.o
[  6%] Generating python/build/lib/ikos/__init__.py
warning: build_py: byte-compiling is disabled, skipping.

[  6%] Built target ikos-python
[  8%] Building CXX object frontend/llvm/CMakeFiles/ikos-pp-lib.dir/src/pass/name_values.cpp.o
[  8%] Building CXX object frontend/llvm/CMakeFiles/ikos-pp-lib.dir/src/pass/remove_printf_calls.cpp.o
[  9%] Building CXX object frontend/llvm/CMakeFiles/ikos-pp-lib.dir/src/pass/remove_unreachable_blocks.cpp.o
[ 11%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/pass/add_loop_counters.cpp.o
[ 11%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/pass/add_partitioning_variables.cpp.o
[ 13%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/pass/name_values.cpp.o
[ 13%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/pass/pass.cpp.o
[ 14%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/pass/simplify_cfg.cpp.o
[ 14%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/pass/simplify_upcast_comparison.cpp.o
In file included from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/bundle.hpp:50,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/pass/pass.hpp:46,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/pass/add_partitioning_variables.hpp:46,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/src/pass/add_partitioning_variables.cpp:50:
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:111:8: error: 'vector' in namespace 'std' does not name a template type
  111 |   std::vector< DataLayoutInfo > integers;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:52:1: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
   51 | #include <ikos/ar/support/number.hpp>
  +++ |+#include <vector>
   52 | 
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:8: error: 'vector' in namespace 'std' does not name a template type
  114 |   std::vector< DataLayoutInfo > floats;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:3: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
  114 |   std::vector< DataLayoutInfo > floats;
      |   ^~~
[ 16%] Building CXX object ar/CMakeFiles/ikos-ar.dir/src/semantic/bundle.cpp.o
In file included from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/bundle.hpp:50,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/pass/pass.hpp:46,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/src/pass/pass.cpp:44:
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:111:8: error: 'vector' in namespace 'std' does not name a template type
  111 |   std::vector< DataLayoutInfo > integers;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:52:1: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
   51 | #include <ikos/ar/support/number.hpp>
  +++ |+#include <vector>
   52 | 
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:8: error: 'vector' in namespace 'std' does not name a template type
  114 |   std::vector< DataLayoutInfo > floats;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:3: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
  114 |   std::vector< DataLayoutInfo > floats;
      |   ^~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/src/pass/add_partitioning_variables.cpp:59:50: error: variable 'constexpr const std::array<unsigned int, 4> ikos::ar::bit_widths' has initializer but incomplete type
   59 | constexpr static const std::array< unsigned, 4 > bit_widths = {1, 8, 32, 64};
      |                                                  ^~~~~~~~~~
In file included from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/bundle.hpp:50,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/pass/pass.hpp:46,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/pass/simplify_cfg.hpp:46,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/src/pass/simplify_cfg.cpp:44:
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:111:8: error: 'vector' in namespace 'std' does not name a template type
  111 |   std::vector< DataLayoutInfo > integers;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:52:1: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
   51 | #include <ikos/ar/support/number.hpp>
  +++ |+#include <vector>
   52 | 
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:8: error: 'vector' in namespace 'std' does not name a template type
  114 |   std::vector< DataLayoutInfo > floats;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:3: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
  114 |   std::vector< DataLayoutInfo > floats;
      |   ^~~
[ 16%] Linking CXX static library libikos-pp.a
[ 16%] Built target ikos-pp-lib
[ 16%] Building CXX object frontend/llvm/CMakeFiles/ikos-pp.dir/src/ikos_pp.cpp.o
make[2]: *** [ar/CMakeFiles/ikos-ar.dir/build.make:132: ar/CMakeFiles/ikos-ar.dir/src/pass/add_partitioning_variables.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [ar/CMakeFiles/ikos-ar.dir/build.make:160: ar/CMakeFiles/ikos-ar.dir/src/pass/pass.cpp.o] Error 1
In file included from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/bundle.hpp:50,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/src/semantic/bundle.cpp:44:
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:111:8: error: 'vector' in namespace 'std' does not name a template type
  111 |   std::vector< DataLayoutInfo > integers;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:52:1: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
   51 | #include <ikos/ar/support/number.hpp>
  +++ |+#include <vector>
   52 | 
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:8: error: 'vector' in namespace 'std' does not name a template type
  114 |   std::vector< DataLayoutInfo > floats;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:3: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
  114 |   std::vector< DataLayoutInfo > floats;
      |   ^~~
make[2]: *** [ar/CMakeFiles/ikos-ar.dir/build.make:174: ar/CMakeFiles/ikos-ar.dir/src/pass/simplify_cfg.cpp.o] Error 1
In file included from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/bundle.hpp:50,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/pass/pass.hpp:46,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/pass/simplify_upcast_comparison.hpp:47,
                 from /tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/src/pass/simplify_upcast_comparison.cpp:44:
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:111:8: error: 'vector' in namespace 'std' does not name a template type
  111 |   std::vector< DataLayoutInfo > integers;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:52:1: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
   51 | #include <ikos/ar/support/number.hpp>
  +++ |+#include <vector>
   52 | 
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:8: error: 'vector' in namespace 'std' does not name a template type
  114 |   std::vector< DataLayoutInfo > floats;
      |        ^~~~~~
/tmp/ikos-20230309-927908-oltumh/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:3: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
  114 |   std::vector< DataLayoutInfo > floats;
      |   ^~~
make[2]: *** [ar/CMakeFiles/ikos-ar.dir/build.make:202: ar/CMakeFiles/ikos-ar.dir/src/semantic/bundle.cpp.o] Error 1
make[2]: *** [ar/CMakeFiles/ikos-ar.dir/build.make:188: ar/CMakeFiles/ikos-ar.dir/src/pass/simplify_upcast_comparison.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1528: ar/CMakeFiles/ikos-ar.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 16%] Linking CXX executable ikos-pp
[ 16%] Built target ikos-pp
make: *** [Makefile:146: all] Error 2

HOMEBREW_VERSION: 4.0.5
ORIGIN: https://github.com/Homebrew/brew
HEAD: b1ef41c25d12795dc93975e337e8644d0974b53c
Last commit: 4 days ago
Core tap JSON: 09 Mar 10:09 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_BOOTSNAP: set
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: octa-core 64-bit haswell
Clang: 10.0.0
Git: 2.25.1 => /bin/git
Curl: 7.68.0 => /bin/curl
Kernel: Linux 5.15.0-60-generic x86_64 GNU/Linux
OS: Linux Mint 20.3 (una)
Host glibc: 2.31
/usr/bin/gcc: 9.4.0
/usr/bin/ruby: 2.7.0
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A

HOMEBREW_CC: gcc-12
HOMEBREW_CXX: g++-12
MAKEFLAGS: -j8

```

**Comment by notnotnull** [comment]
Also getting this issue. Did you forget to #include \<vector\>?

**Comment by mschwerhoff** [comment]
Similar situation (seems related to issues #5 and #6):

```
$ wsl.exe -l -v
  NAME                   STATE           VERSION
* Ubuntu-20.04           Running         2
  ...
```

```
$ uname -a
Linux streamer 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`
```

```
$ brew install nasa-sw-vnv/core/ikos
...
==> Installing nasa-sw-vnv/core/ikos
==> python3 -m venv --system-site-packages /home/linuxbrew/.linuxbrew/Cellar/ikos/3.1_1/libexec/vendor
==> /home/linuxbrew/.linuxbrew/Cellar/ikos/3.1_1/libexec/vendor/bin/pip install -v --no-deps --no-binary :all: --use-feature=no-binary-enable-wheel-cache --i
==> cmake -G Unix Makefiles -DGMP_ROOT=/home/linuxbrew/.linuxbrew/opt/gmp -DMPFR_ROOT=/home/linuxbrew/.linuxbrew/opt/mpfr -DPPL_ROOT=/home/linuxbrew/.linuxbr
==> make install
Last 15 lines from /home/malte/.cache/Homebrew/Logs/ikos/04.make:
   51 | #include <ikos/ar/support/number.hpp>
  +++ |+#include <vector>
   52 |
/tmp/ikos-20230530-10071-etz05b/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:8: error: ‘vector’ in namespace ‘std’ does not name a template type
  114 |   std::vector< DataLayoutInfo > floats;
      |        ^~~~~~
/tmp/ikos-20230530-10071-etz05b/ikos-3.1/ar/include/ikos/ar/semantic/data_layout.hpp:114:3: note: ‘std::vector’ is defined in header ‘<vector>’; did you forget to ‘#include <vector>’?
  114 |   std::vector< DataLayoutInfo > floats;
      |   ^~~
make[2]: *** [ar/CMakeFiles/ikos-ar.dir/build.make:160: ar/CMakeFiles/ikos-ar.dir/src/pass/pass.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1497: ar/CMakeFiles/ikos-ar.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 12%] Linking CXX static library libikos-pp.a
[ 12%] Built target ikos-pp-lib
make: *** [Makefile:146: all] Error 2

If reporting this issue please do so at (not Homebrew/brew or Homebrew/homebrew-core):
  https://github.com/nasa-sw-vnv/homebrew-core/issues

These open issues may also help:
IKOS build fails on gcc12 https://github.com/NASA-SW-VnV/homebrew-core/issues/6
Failed to install IKOS v3.1 with Homebrew on Ubuntu 20.04 (WSL-2) https://github.com/NASA-SW-VnV/homebrew-core/issues/5
```

**Comment by jsforyu** [comment]
Also having this problem .Have you already resolved it？

**Comment by PhilipBotha** [comment]
Same issue with GCC Version 11.4.0 as well on WSL 2 Ubuntu 22.04

**Comment by ivanperez-keera** [comment]
I believe this issue is related to: https://github.com/NASA-SW-VnV/ikos/pull/201

If so, we've just announced a release candidate for IKOS 3.2, which includes that fix. We've included instructions to install it with homebrew. See https://github.com/NASA-SW-VnV/ikos/discussions/253 for details.

Please try it and let us know if the issue persists. Thanks!

**Comment by ivanperez-keera** [comment]
Closed by the solution to #9.
